software/displayerang.livecodescript

```diff
@@ -0,0 +1,1402 @@
+               filter pData["abutDive"][tRec] without empty
+            end if
+         end repeat
+      end repeat
+   end repeat
+   -- find the case node
+   -- needed to request certain changes
+   repeat for each key tRec in pData["nodes"]
+      if "bulk" is among the keys of pData["nodes"][tRec] and "read" is among the keys of pData["nodes"][tRec] then
+         put tRec into pData["case"]
+      end if
+   end repeat
+   -- undo/redo
+   put pShort[pDoc][0]["undo"] into pData["undo"]
+   put pShort[pDoc][0]["redo"] into pData["redo"]
+   -- metatdata
+   put pShort[pDoc][0]["pointerTags"] into pData["pointerTags"]
+   -- node display label ----------------
+   repeat for each key tRec in pData["nodes"]
+      put pShort[pDoc][tRec]["title"] into tForShow
+      if tForShow is empty then
+         --         put symbolLookup(pShort[pDoc][tRec]["type"]) && tRec into tForShow
+         put "<p>" & symbolLookupHTML(pShort[pDoc][tRec]["type"]) && tRec & "</p>" into tForShow
+      else
+         --         put symbolLookup(pShort[pDoc][tRec]["type"]) & " " before tForShow
+         put symbolLookupHTML(pShort[pDoc][tRec]["type"]) into tSymbol
+         put "<p>" & tSymbol & space before tForShow
+         put "</p>" after tForShow
+      end if
+      put tForShow into pData["nodes"][tRec]["displayString"]
+      put tForShow into pData["nodes"][tRec]["displayStringExtended"]
+      -- pass forward the memo & figure if they're present
+      put pShort[pDoc][tRec]["memo"] into tForShow
+      if tForShow is not empty then put cr & "<p>" & tForShow & "</p>" after pData["nodes"][tRec]["displayStringExtended"]
+      put pShort[pDoc][tRec]["figure"] into tForShow
+      -- set all of the <img> tags to the same image object
+      -- later, if this figure needs to be shown in a field (like tooltip), that image object's filename will be set to the figure's path
+      -- the field's image won't care what's in the image object
+      -- 17624 is the ID of image "figureSource"
+      if tForShow is not empty then 
+         put cr & "<p><img src=17624></p>" after pData["nodes"][tRec]["displayStringExtended"]
+         
+      end if
+   end repeat
+end tempConvertData
+
+on middleOut pData,@pDraw,@pFocus -- ! a hacky way to draw the provided graph, calls assignLayers
+   -- ! as of 10JAN2016 middleOut only follows flow links, so it won't visit or follow dive/rise links
+   put "on middleOut" into xX["write"];    devLog xX
+   -- pData ! just sGraphData at the moment, maybe this should be an adjacency list ?
+   -- setup variables --------------------------------------------------------------------------------
+   put empty into pData["layers"]
+   put empty into pData["visited"]
+   -- make sure the focus node is actually focusable
+   put the keys of pData["nodes"] into tKeys
+   put empty into tRecs
+   repeat for each line tLine in tKeys
+      if tLine is a number then put tLine into line (the number of lines of tRecs)+1 of tRecs
```

Comment:
@matthewmaier Putting tLine & return after tRecs would be faster because it is optimised and also it would not need to count the lines each loop. You will just need to handle the trailing return